### PR TITLE
[WIP] Greenplum backup push

### PR DESCRIPTION
**Pull request by nastprol**

### CODE REVIEW

**Comment by usernamedt** [comment]
Hi!
I am having trouble running `backup-push` on Greenplum 6.15:
```
WARNING: 2021/06/07 15:53:56.745012 Couldn't get system identifier because of error: 'System Identifier: getting identifier of DB failed: ERROR: function pg_control_system() does not exist (SQLSTATE 42883)'
INFO: 2021/06/07 15:53:56.749770 Running wal-g on segments
panic: 20210607:15:53:56 wal-g:root:xxxxx:079003-[CRITICAL]:-Unable to run wal-g on 5 segments. See /root/gpAdminLogs/wal-g_20210607.log for a complete list of errors.

goroutine 1 [running]:
github.com/greenplum-db/gp-common-go-libs/gplog.abort(0xc000a7f950, 0x1, 0x1)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:353 +0x105
github.com/greenplum-db/gp-common-go-libs/gplog.Fatal(0x1d54d00, 0xc0001a2d00, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:303 +0x465
github.com/greenplum-db/gp-common-go-libs/cluster.LogFatalClusterError(0xc0001ae320, 0x16, 0xc000a7fa02, 0x5)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/cluster/cluster.go:398 +0x21a
github.com/greenplum-db/gp-common-go-libs/cluster.(*Cluster).CheckClusterError(0xc0003fba40, 0xc000690580, 0x1b1e7be, 0x13, 0x17717a0, 0x1b73a28, 0x0, 0x0, 0x0)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/cluster/cluster.go:380 +0x565
github.com/wal-g/wal-g/internal/databases/greenplum.(*BackupHandler).HandleBackupPush(0xc0003f8480)
	/home/usernamedt/gp_walg/wal-g/internal/databases/greenplum/backup_push_handler.go:103 +0x24f
github.com/wal-g/wal-g/cmd/gp.glob..func1(0x2ced800, 0xc00000fe60, 0x0, 0x2)
	/home/usernamedt/gp_walg/wal-g/cmd/gp/backup_push.go:66 +0x225
github.com/spf13/cobra.(*Command).execute(0x2ced800, 0xc00000fde0, 0x2, 0x2, 0x2ced800, 0xc00000fde0)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:830 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x2ceda80, 0x45302a, 0x2d065e0, 0xc000000180)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:864
github.com/wal-g/wal-g/cmd/gp.Execute()
	/home/usernamedt/gp_walg/wal-g/cmd/gp/gp.go:33 +0x31
main.main()
	/home/usernamedt/gp_walg/wal-g/main/gp/main.go:8 +0x25
```

It would be cool to create some basic docker test to make sure that WAL-G can create a backup without errors.

**Comment by nastprol** [comment]
> Hi!
> I am having trouble running `backup-push` on Greenplum 6.15:
> 
> ```
> WARNING: 2021/06/07 15:53:56.745012 Couldn't get system identifier because of error: 'System Identifier: getting identifier of DB failed: ERROR: function pg_control_system() does not exist (SQLSTATE 42883)'
> INFO: 2021/06/07 15:53:56.749770 Running wal-g on segments
> panic: 20210607:15:53:56 wal-g:root:xxxxx:079003-[CRITICAL]:-Unable to run wal-g on 5 segments. See /root/gpAdminLogs/wal-g_20210607.log for a complete list of errors.
> 
> goroutine 1 [running]:
> github.com/greenplum-db/gp-common-go-libs/gplog.abort(0xc000a7f950, 0x1, 0x1)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:353 +0x105
> github.com/greenplum-db/gp-common-go-libs/gplog.Fatal(0x1d54d00, 0xc0001a2d00, 0x0, 0x0, 0x0, 0x0, 0x0)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/gplog/gplog.go:303 +0x465
> github.com/greenplum-db/gp-common-go-libs/cluster.LogFatalClusterError(0xc0001ae320, 0x16, 0xc000a7fa02, 0x5)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/cluster/cluster.go:398 +0x21a
> github.com/greenplum-db/gp-common-go-libs/cluster.(*Cluster).CheckClusterError(0xc0003fba40, 0xc000690580, 0x1b1e7be, 0x13, 0x17717a0, 0x1b73a28, 0x0, 0x0, 0x0)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/greenplum-db/gp-common-go-libs/cluster/cluster.go:380 +0x565
> github.com/wal-g/wal-g/internal/databases/greenplum.(*BackupHandler).HandleBackupPush(0xc0003f8480)
> 	/home/usernamedt/gp_walg/wal-g/internal/databases/greenplum/backup_push_handler.go:103 +0x24f
> github.com/wal-g/wal-g/cmd/gp.glob..func1(0x2ced800, 0xc00000fe60, 0x0, 0x2)
> 	/home/usernamedt/gp_walg/wal-g/cmd/gp/backup_push.go:66 +0x225
> github.com/spf13/cobra.(*Command).execute(0x2ced800, 0xc00000fde0, 0x2, 0x2, 0x2ced800, 0xc00000fde0)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:830 +0x2c2
> github.com/spf13/cobra.(*Command).ExecuteC(0x2ceda80, 0x45302a, 0x2d065e0, 0xc000000180)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:914 +0x30b
> github.com/spf13/cobra.(*Command).Execute(...)
> 	/home/usernamedt/gp_walg/wal-g/vendor/github.com/spf13/cobra/command.go:864
> github.com/wal-g/wal-g/cmd/gp.Execute()
> 	/home/usernamedt/gp_walg/wal-g/cmd/gp/gp.go:33 +0x31
> main.main()
> 	/home/usernamedt/gp_walg/wal-g/main/gp/main.go:8 +0x25
> ```
> 
> It would be cool to create some basic docker test to make sure that WAL-G can create a backup without errors.

Hi!
I added the test as you requested. You can run `make gp_test`

**Comment by usernamedt** [comment]
> Hi!
> I added the test as you requested. You can run `make gp_test`

This needs to be included into the [CI workflow](https://github.com/wal-g/wal-g/blob/490b27137551f1dbccac4a61ffe27f728e976ad4/.github/workflows/dockertests.yml#L73).

File: internal/databases/postgres/connect.go [hunk 1]
```diff
@@ -26,7 +26,15 @@ func Connect(configOptions ...func(config *pgx.ConnConfig) error) (*pgx.Conn, er
 	}
 	conn, err := pgx.Connect(config)
 	if err != nil {
-		if config.Host != "localhost" {
+		config.RuntimeParams["gp_role"] = "utility"
```
Comment (usernamedt) [inline internal/databases/postgres/connect.go:29]: I think it is OK to leave this hinge at the MVP stage, but I suggest extracting this logic to the explicit function to make it clear what is going on. For example, you may name it `tryConnectToGpSegment` or something like that. And leave some comment regarding this logic.

File: internal/databases/postgres/queryRunner.go [hunk 2]
```diff
@@ -385,3 +395,109 @@ func (queryRunner *PgQueryRunner) GetPhysicalSlotInfo(slotName string) (Physical
 func (queryRunner *PgQueryRunner) IsTablespaceMapExists() bool {
 	return queryRunner.Version >= 90600
 }
+
+// BuildCreateGreenplumRestorePoint formats a query to create a restore point for Greenplum
+func (queryRunner *PgQueryRunner) buildCreateGreenplumRestorePoint(restorePointName string) string {
```
Comment (usernamedt) [inline internal/databases/postgres/queryRunner.go:400]: Better move the Greenplum-specific functions to the separate query runner.
You can introduce the GpQueryRunner and embed the existing PgQueryRunner into it.

File: internal/databases/postgres/timeline.go [hunk 3]
```diff
@@ -162,7 +163,8 @@ func ParseTimelineFromBackupName(backupName string) (uint32, error) {
 		return 0, newIncorrectBackupNameError(backupName)
```
Comment (usernamedt) [inline internal/databases/postgres/timeline.go:163]: Add a couple of unit tests for new functionality regarding the possible backup name prefix.

**Review by usernamedt — CHANGES_REQUESTED** [review]
need a couple of fixes before merging

File: docker/pg_tests/scripts/tests/backup_add_name_prefix_test.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+wal-g --config=${TMP_CONFIG} delete everything FORCE --confirm
+
+pgbench -i -s 5 postgres
+pg_dumpall -f /tmp/dump1
+pgbench -c 2 -T 100000000 -S &
+sleep 1
+wal-g --config=${TMP_CONFIG} backup-push ${PGDATA} --backup-name-prefix prefix
+
+wal-g --config=${TMP_CONFIG} backup-list
```
Comment (usernamedt) [inline docker/pg_tests/scripts/tests/backup_add_name_prefix_test.sh:28]: So, as I understood, the logic is to check the name of the resulting backup after calling the backup-list? If so, I guess this test lacks a couple of ending lines :)

File: internal/databases/greenplum/queryRunner.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	"github.com/wal-g/wal-g/internal/databases/postgres"
+)
+
+// GpQueryRunner is implementation for controlling Greenplum
+type GpQueryRunner struct {
+	pgQueryRunner *postgres.PgQueryRunner
+}
+
+// NewPgQueryRunner builds QueryRunner from available connection
+func NewPgQueryRunner(conn *pgx.Conn) (*GpQueryRunner, error) {
```
Comment (usernamedt) [inline internal/databases/greenplum/queryRunner.go:20]: `NewGpQueryRunner`

**Review by usernamedt — CHANGES_REQUESTED** [review]

**Review by usernamedt — APPROVED** [review]

**Pull request merged**